File: wot-dijkstra/src/main/java/org/pgpainless/wot/dijkstra/sq/Certification.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq;

import java.util.Date;

public class Certification {

    private final CertSynopsis issuer;
    private final CertSynopsis target;
    private final Optional<String> userId;

    private final Date creationTime;
    private final Optional<Date> expirationTime;
    private final boolean exportable;
    private final int trustAmount;
    private final Depth trustDepth;
    private final RegexSet regex;

    public Certification(
            CertSynopsis issuer,
            CertSynopsis target,
            Optional<String> userId,
            Date creationTime,
            Optional<Date> expirationTime,
            boolean exportable,
            int trustAmount,
            Depth trustDepth,
            RegexSet regex) {
        this.issuer = issuer;
        this.target = target;
        this.userId = userId;
        this.creationTime = creationTime;
        this.expirationTime = expirationTime;
        this.exportable = exportable;
        this.trustAmount = trustAmount;
        this.trustDepth = trustDepth;
        this.regex = regex;
    }

    public Certification(CertSynopsis issuer,
                         Optional<String> targetUserId,
                         CertSynopsis target,
                         Date creationTime) {
        this.issuer = issuer;
        this.target = target;
        this.userId = targetUserId;
        this.creationTime = creationTime;

        this.expirationTime = Optional.empty();
        this.exportable = true;
        this.trustDepth = Depth.limited(0);
        this.trustAmount = 120;
        this.regex = RegexSet.wildcard();
    }

    /**
     * Get the issuer of the certification.
     *
     * @return issuer
     */
    public CertSynopsis getIssuer() {
        return issuer;
    }

    /**
     * Get the target of the certification.
     *
     * @return target
     */
    public CertSynopsis getTarget() {
        return target;
    }

    /**
     * Get the target user-id.
     *
     * @return user-id
     */
    public Optional<String> getUserId() {
        return userId;
    }

    /**
     * Get the creation time of the certification.
     *
     * @return creation time
     */
    public Date getCreationTime() {
        return creationTime;
    }

    /**
     * Get the (optional) expiration time of the certification.
     *
     * @return optional expiration time
     */
    public Optional<Date> getExpirationTime() {
        return expirationTime;
    }

    /**
     * Return true if the certification is marked as exportable.
     *
     * @return exportable
     */
    public boolean isExportable() {
        return exportable;
    }

    /**
     * Get the trust amount of the certification.
     *
     * @return trust amount
     */
    public int getTrustAmount() {
        return trustAmount;
    }

    /**
     * Get the trust depth of the certification.
     *
     * @return trust depth
     */
    public Depth getTrustDepth() {
        return trustDepth;
    }

    /**
     * Return the set of regular expressions.
     *
     * @return regex set
     */
    public RegexSet getRegexes() {
        return regex;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(issuer.getFingerprint()).append((issuer.userIds().isEmpty() ? " " : " (" + issuer.userIds().iterator().next() + ") "));
        sb.append(userId.isPresent() ? "certifies" : "delegates to").append(userId.isPresent() ? " [" + userId.get() + "] " : " ").append(target.getFingerprint())
                .append(userId.isEmpty() && !target.userIds().isEmpty() ? " (" + target.userIds().iterator().next() + ")" : "");
        return sb.toString();
    }
}

File: wot-dijkstra/src/main/kotlin/org/pgpainless/wot/dijkstra/sq/Certification.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq

import java.util.*

/**
 * A [Certification] is a signature issued by one certificate over a datum on another target certificate.
 * Such a datum could be either a user-id, or the primary key of the target certificate.
 *
 * @param issuer synopsis of the certificate that issued the [Certification]
 * @param target synopsis of the certificate that is target of this [Certification]
 * @param userId optional user-id. If this is null, the [Certification] is made over the primary key of the target.
 * @param creationTime creation time of the [Certification]
 * @param expirationTime optional expiration time of the [Certification]
 * @param exportable if false, the certification is marked as "not exportable"
 * @param trustAmount amount of trust the issuer places in the binding
 * @param trustDepth degree to which the issuer trusts the target as trusted introducer
 * @param regexes regular expressions for user-ids which the target is allowed to introduce
 */
data class Certification(
        val issuer: CertSynopsis,
        val target: CertSynopsis,
        val userId: String?,
        val creationTime: Date,
        val expirationTime: Date?,
        val exportable: Boolean,
        val trustAmount: Int,
        val trustDepth: Depth,
        val regexes: RegexSet
) {

    /**
     * Construct a [Certification] with default values. The result is non-expiring, will be exportable and has a
     * trust amount of 120, a depth of 0 and a wildcard regex.
     *
     * @param issuer synopsis of the certificate that issued the [Certification]
     * @param target synopsis of the certificate that is target of this [Certification]
     * @param targetUserId optional user-id. If this is null, the [Certification] is made over the primary key of the target.
     * @param creationTime creation time of the [Certification]
     */
    constructor(
            issuer: CertSynopsis,
            targetUserId: String?,
            target: CertSynopsis,
            creationTime: Date) :
            this(issuer, target, targetUserId, creationTime, null, true, 120, Depth.limited(0), RegexSet.wildcard())

    override fun toString(): String {
        return if (userId != null)
            "$issuer certifies [$userId] ${target.fingerprint}"
        else
            "$issuer delegates to ${target.fingerprint}"
    }
}